### PR TITLE
To support fd passing

### PR DESCRIPTION
**Pull request by luckyh**
aexpect.client: To support fd passing
    
Add new argument `pass_fds` for `Spawn`/`Tail`/`Expect`/`ShellSession`
and their related wrappers to support fd passing.
    
In python 2, the above utilities do support (implicit) fd passing, but
these would not work in python 3 for some reasons. By applying this
patch, now users can pass fds to the target process explicitly, and
implicit fd passing is not allowed.
    
Signed-off-by: Xu Han <xuhan@redhat.com>

### CODE REVIEW

**Comment by luckyh** [comment]
test_aexpect.py:
```python
import aexpect

f = open("/dev/null", "r")
p = aexpect.run_tail("./show_fd.py")
print(p.get_output())
p = aexpect.run_tail("./show_fd.py", pass_fds=[f.fileno()])
print(p.get_output())
```

show_fd.py:
```python
#!/usr/bin/python

import os
os.system("ls -l /proc/%s/fd" % os.getpid())
```

Test results:

python 2:
```
python2 test_aexpect.py 
total 0
lrwx------. 1 root root 64 May 21 15:45 0 -> /dev/pts/1
lrwx------. 1 root root 64 May 21 15:45 1 -> /dev/pts/1
lrwx------. 1 root root 64 May 21 15:45 2 -> /dev/pts/1
l-wx------. 1 root root 64 May 21 15:45 3 -> /tmp/aexpect_WUOQUOv6/server-log

total 0
lrwx------. 1 root root 64 May 21 15:45 0 -> /dev/pts/1
lrwx------. 1 root root 64 May 21 15:45 1 -> /dev/pts/1
lrwx------. 1 root root 64 May 21 15:45 2 -> /dev/pts/1
lr-x------. 1 root root 64 May 21 15:45 3 -> /dev/null
lrwx------. 1 root root 64 May 21 15:45 4 -> /tmp/aexpect_mwpt6WsS/lock-client-starting
lr-x------. 1 root root 64 May 21 15:45 5 -> /tmp/aexpect_WUOQUOv6/outpipe-tail
l-wx------. 1 root root 64 May 21 15:45 6 -> /tmp/aexpect_mwpt6WsS/server-log

```

python 3:
```
$ python3 test_aexpect.py 
total 0
lrwx------. 1 root root 64 May 21 15:42 0 -> /dev/pts/1
lrwx------. 1 root root 64 May 21 15:42 1 -> /dev/pts/1
lrwx------. 1 root root 64 May 21 15:42 2 -> /dev/pts/1

total 0
lrwx------. 1 root root 64 May 21 15:42 0 -> /dev/pts/1
lrwx------. 1 root root 64 May 21 15:42 1 -> /dev/pts/1
lrwx------. 1 root root 64 May 21 15:42 2 -> /dev/pts/1
lr-x------. 1 root root 64 May 21 15:42 3 -> /dev/null

```

File: aexpect/utils/process.py [hunk 1]
```diff
@@ -102,3 +104,32 @@ def process_in_ptree_is_defunct(ppid):
...
             break
     return defunct
+
+
+PIPE = subprocess.PIPE
+STDOUT = subprocess.STDOUT
+
+
+if sys.version_info < (3, 2):
+    # Actually we should rewrite `subprocess.Popen` with a complete
```
Comment (ldoktor) [inline aexpect/utils/process.py]: Well actually we can just use `subprocess32` backport and set the `fcntl` there (either as suggested here, or by using `subprocess._set_cloexec(fd, False)`, but as that one is private I'd suggest the `fcntl`.

File: aexpect/client.py [hunk 2]
```diff
@@ -1378,12 +1396,15 @@ def run_fg(command, output_func=None, output_prefix="", timeout=1.0):
...
+    :param pass_fds: Optional sequence of file descriptors to keep open
+            between the parent and child.
 
     :return: A 2-tuple containing the exit status of the process and its
             STDOUT/STDERR output.  If timeout expires before the process
             terminates, the returned status is None.
     """
-    bg_process = run_bg(command, None, output_func, output_prefix, timeout)
+    bg_process = run_bg(command, None, output_func, output_prefix, timeout,
+                        pass_fds=pass_fds)
```
Comment (ldoktor) [inline aexpect/client.py]: yep, this commit looks good.

**Comment by ldoktor** [comment]
I see so without this change on py2 even the first process gets the `/dev/null` fd and on py3 no process get's the `/dev/null`, while with those fixes first execution masks the `/dev/null` and second passes it to both py2/py3.

Dangerous thing is that some people might depend on the old behavior so I'm not sure we can require the `pass_fds` on py2. On the other hand the usage of such feature is quite limited so perhaps we can try to educate people and require that, unless it breaks someones setup too badly.

So I like this approach, the only thing I'd change is the use of `subprocess32` on py2 instead of custom behavior (with the `fcntl` to mask the `FD_CLOEXEC`) as that one is a standard interface and could save us some troubles.

It'd be also nice if you could add the selftest along with check that `/dev/null` is handled properly. It doesn't have to be fancy, just use `in` or `not in` the output. There are currently no tests, but in https://github.com/avocado-framework/aexpect/pull/28 they are about to be introduced. You can simply create this test in `tests/test_pass_fds.py` and not care about the execution for now.

**Comment by luckyh** [comment]
@ldoktor created https://github.com/avocado-framework/aexpect/pull/33 per your suggestion, thanks.

**Pull request closed without merging**